File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py

def solution(numbers, target):
    answer = 0
    def dfs(nums_sum, index):
        if index == len(numbers):
            if nums_sum == target:
                nonlocal answer
                answer += 1
            return
        dfs(nums_sum + numbers[index], index + 1)
        dfs(nums_sum - numbers[index], index + 1)
    dfs(0, 0)
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int answer = 0;
    
    public int solution(int[] numbers, int target) {
        dfs(numbers, 0, target, 0);
        return answer;
    }
    
    public void dfs(int[] numbers, int index, int target, int sum){
        //종료 조건
        if(index == numbers.length){
            if(sum == target){
                answer++;
            }
            return;
        }
        
        dfs(numbers, index + 1, target, sum + numbers[index]);
        dfs(numbers, index + 1, target, sum - numbers[index]);
    }
    
    
}